mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,689 @@
+                                    }
+                                },
+                                None => break Ok(()), // PortForwarder received DaemonTcpOutgoing::Close
+                            }
+                        }
+                    }
+                });
+            }
+            (socket, Err(error)) => {
+                // error from TcpStream
+                tracing::error!("Error occured while listening to local socket {socket}: {error}");
+                self.listeners.remove(&socket);
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_msg_from_task(
+        &mut self,
+        message: PortForwardMessage,
+    ) -> Result<(), PortForwardError> {
+        match message {
+            PortForwardMessage::Connect(remote_socket, oneshot) => {
+                let remote_address = SocketAddress::Ip(remote_socket);
+                self.oneshots.push_back(oneshot);
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Connect(
+                        LayerConnect { remote_address },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "Failed to send ClientMessage to agent".into(),
+                    ));
+                };
+            }
+            PortForwardMessage::Send(connection_id, bytes) => {
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Write(
+                        LayerWrite {
+                            connection_id,
+                            bytes,
+                        },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "Failed to send ClientMessage to agent".into(),
+                    ));
+                };
+            }
+            PortForwardMessage::Close(socket, connection_id) => {
+                // end of listener TCP stream in task
+                let Ok(_) = self
+                    .agent_connection
+                    .sender
+                    .send(ClientMessage::TcpOutgoing(LayerTcpOutgoing::Close(
+                        LayerClose { connection_id },
+                    )))
+                    .await
+                else {
+                    return Err(PortForwardError::AgentError(
+                        "Failed to send ClientMessage to agent".into(),
+                    ));
+                };
+                self.tasks.remove(&socket);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum PortForwardError {
+    #[error("Failed to setup connection with Agent with error: `{0}`")]
+    AgentSetupError(String),
```

Comment:
I think you can add this instead:
```rust
pub enum PortForwardError {
    ...
    #[error("connection with the agent failed")]
    AgentConnectionFailed,
    ...
}

impl From<mpsc::error::SendError<ClientMessage>> for PortForwardError {
    fn from(_: mpsc::error::SendError<ClientMessage>) -> Self {
        Self::AgentConnectionFailed
    }
}
```
Use it whenever you fail to send/recv using `agent_connection`